packages/react-dom/src/server/ReactDOMServerExternalRuntime.js

```diff
@@ -1,18 +1,164 @@
-// TODO: Add Flow types
+/**
+ * This file is compiled to a standalone browser script by rollup and loaded by Fizz
+ *  clients. Therefore, it should be fast and not have many external dependencies.
+ * @flow
+ */
+
+// Imports are resolved statically by the closure compiler in release bundles
+// and by rollup in jest unit tests
 import {
   clientRenderBoundary,
   completeBoundaryWithStyles,
   completeBoundary,
   completeSegment,
-} from 'react-dom-bindings/src/server/fizz-instruction-set/ReactDOMFizzInstructionSet';
-
-// Intentionally does nothing. Implementation will be added in future PR.
-// eslint-disable-next-line no-unused-vars
-const observer = new MutationObserver(mutations => {
-  // These are only called so I can check what the module output looks like. The
-  // code is unreachable.
-  clientRenderBoundary();
-  completeBoundaryWithStyles();
-  completeBoundary();
-  completeSegment();
-});
+} from '../../../react-dom-bindings/src/server/fizz-instruction-set/ReactDOMFizzInstructionSet';
+
+// This runtime may be sent to the client multiple times (if FizzServer.render
+//  is called more than once). Here, we check whether the mutation observer
+//  was already created / installed
+if (!window.reactInstructionObserver && window.MutationObserver) {
+  // TODO: eventually remove
+  // We currently need to set these globals for compatibility with
+  // ReactDOMFizzInstructionSet
+  window.$RC = completeBoundary;
+  window.$RM = new Map();
+  window.reactInstructionObserver = new MutationObserver(mutations => {
+    mutations.forEach(mutation => {
+      mutation.addedNodes.forEach(handleNode);
+    });
+  });
+
+  const existingNodes = document.getElementsByTagName('div');
+  for (let i = 0; i < existingNodes.length; i++) {
+    handleNode(existingNodes[i]);
+  }
+
+  window.reactInstructionObserver.observe(document.body, {
+    childList: true,
+    subtree: true,
+  });
+}
+
+function handleNode(node_ /*: Node */) {
+  if (node_.nodeType !== 1) {
+    return;
+  }
+  // $FlowFixMe[incompatible-type]
+  const node /*: HTMLElement*/ = node_;
+  const instr = node.getAttribute('data-react-instr');
+  if (!instr) {
+    return;
+  }
+
+  switch (instr) {
+    case '$RX':
+      clientRenderBoundaryImpl(
+        node.getAttribute('data-rarg0'),
```

Comment:
The way we access these in the other runtime is slightly different:

```js
const dataset = node.data
dataset["rarg0"]
```

See: https://github.com/facebook/react/blob/765805bf88f7e4ac8db7b0a587bd82c7ac404220/packages/react-dom-bindings/src/server/fizz-instruction-set/ReactDOMFizzInstructionSet.js#L36

That would compress slightly better. But also we might as well be consistent.

Also, maybe these should use the $ prefix convention? And instead of `rarg1` make it as short as possible to avoid clashes.

3-4 characters seems like enough, like `$ra1`, `$ra2`, etc.

Reply:
Thanks for pointing out `dataset`, will change my code to use that!

I had originally used $r. However, reading through the html spec for data attribute names ([link](https://html.spec.whatwg.org/multipage/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes)), I think the name may need to be [XML compatible](https://www.w3.org/TR/xml/#NT-Name). If this is the case, that would rule out `$`.

Happy to defer to you here though - I'm not at all familiar with specs + browser implementations

Reply:
Oh right. I remember now that `useId` generates ids with `:` for this reason. The convention for `useId` is `:r1:` so we can do something like that:

`:a0`, `:a1` and so on.

In a follow up, what we should do is support the `identifierPrefix` option. Then prepend that to all of these attributes. We can send down prefix as an attribute early in the stream. So that the external runtime knows which attributes to look for. We wanted to do that anyway so that users don't have to manually pass the prefix to `hydrateRoot`.